src/NumericalAlgorithms/Spectral/Spectral.cpp

```diff
@@ -0,0 +1,306 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <vector>
+
+#include "DataStructures/Matrix.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace Spectral {
+
+namespace {
+
+template <Basis BasisType, Quadrature QuadratureType>
+const detail::CollocationPointsAndWeights<BasisType, QuadratureType>&
+collocation_points_and_weights(const size_t number_of_pts) noexcept {
+  ASSERT(number_of_pts > 1, "Must have at least two collocation points");
+  ASSERT(number_of_pts <= ::Spectral::maximum_number_of_pts,
+         "Exceeded maximum number of collocation points.");
+  static const auto collo_pts_and_weights = []() {
+    std::vector<detail::CollocationPointsAndWeights<BasisType, QuadratureType>>
+        local_collo_pts;
+    local_collo_pts.reserve(::Spectral::maximum_number_of_pts - 1);
+    for (size_t n = 2; n <= Spectral::maximum_number_of_pts; ++n) {
+      local_collo_pts.emplace_back(n);
+    }
+    return local_collo_pts;
+  }();
+  return collo_pts_and_weights[number_of_pts - 2];
```

Comment:
the 2 needs to change if we allow 1 Gauss point; we need to comment why the index is offset from the number_of_pts)